### PR TITLE
[AppService] Fix #13888 : Add support for Static WebApps: get, list, create commands

### LINKED ISSUE #13888: webapp: Add support for Static WebApps
**Is your feature request related to a problem? Please describe.**

**Describe the solution you'd like**

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by ElleTojaroon**
**Description**  
Add new commands for new public preview resource type: Static WebApps. Fixes #13888

**Testing Guide**  

TBD

---

This checklist is used to make sure that common guidelines for a pull request are followed.

- [x] The PR title and description has followed the guideline in [Submitting Pull Requests](https://github.com/Azure/azure-cli/tree/dev/doc/authoring_command_modules#submitting-pull-requests).

- [x] I adhere to the [Command Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md).

### CODE REVIEW

**Comment by yonzhan** [comment]
appservice

File: src/azure-cli/azure/cli/command_modules/appservice/_help.py [hunk 1]
```diff
@@ -1991,3 +1991,75 @@
     - name: Delete app service environment.
       text: az appservice ase delete -n MyAseName
 """
+
+helps['staticapp list'] = """
+    type: command
+    short-summary: List all static app resources in a subscription, or in resource group if provided
+    examples:
+    - name: List static apps.
```
Comment (qwordy) [inline src/azure-cli/azure/cli/command_modules/appservice/_help.py:1999]: add "in subscription"

File: src/azure-cli/azure/cli/command_modules/appservice/_help.py [hunk 2]
```diff
@@ -1991,3 +1991,140 @@
     - name: Delete app service environment.
       text: az appservice ase delete -n MyAseName
 """
+
+helps['staticapp list'] = """
```
Comment (qwordy) [inline src/azure-cli/azure/cli/command_modules/appservice/_help.py:1995]: Could you add an entry of "staticapp"?

File: src/azure-cli/azure/cli/command_modules/appservice/_help.py [hunk 3]
```diff
@@ -1991,3 +1991,140 @@
...
+    examples:
+    - name: Connect a repo and branch to static app.
+      text: az staticapp reconnect -n MyStaticAppName --source MyGitHubRepo -b master --token MyAccessToken
+"""
+
+helps['staticapp delete'] = """
+    type: command
+    short-summary: Delete a static app.
+    examples:
+    - name: Delete static app.
```
Comment (qwordy) [inline src/azure-cli/azure/cli/command_modules/appservice/_help.py:2039]:
```suggestion
    - name: Delete a static app.
```

File: src/azure-cli/azure/cli/command_modules/appservice/_help.py [hunk 4]
```diff
@@ -1991,3 +1991,140 @@
...
+      text: az staticapp delete -n MyStaticAppName -g MyRg
+"""
+
+helps['staticapp environments list'] = """
+    type: command
+    short-summary: List all environments of the static app including production.
+    examples:
+    - name: List static app environment.
```
Comment (qwordy) [inline src/azure-cli/azure/cli/command_modules/appservice/_help.py:2047]:
```suggestion
    - name: List static app environments.
```

File: src/azure-cli/azure/cli/command_modules/appservice/_help.py [hunk 5]
```diff
@@ -1991,3 +1991,140 @@
...
+      text: az staticapp environments list -n MyStaticAppName
+"""
+
+helps['staticapp environments show'] = """
+    type: command
+    short-summary: Show information about the production environment or the specified environment.
+    examples:
+    - name: Show static app environment.
```
Comment (qwordy) [inline src/azure-cli/azure/cli/command_modules/appservice/_help.py:2055]:
```suggestion
    - name: Show a static app environment.
```

File: src/azure-cli/azure/cli/command_modules/appservice/_help.py [hunk 6]
```diff
@@ -1991,3 +1991,140 @@
...
+    examples:
+    - name: Show static app.
+      text: az staticapp show -n MyStaticAppName
+"""
+
+helps['staticapp create'] = """
+    type: command
+    short-summary: Create a static app and returns the app created.
+    examples:
+    - name: Create static app.
```
Comment (qwordy) [inline src/azure-cli/azure/cli/command_modules/appservice/_help.py:2015]: Maybe you can add more description about `-s MyGitHubRepo -l Location`

File: src/azure-cli/azure/cli/command_modules/appservice/_help.py [hunk 7]
```diff
@@ -1991,3 +1991,140 @@
...
+
+helps['staticapp delete'] = """
+    type: command
+    short-summary: Delete a static app.
+    examples:
+    - name: Delete static app.
+      text: az staticapp delete -n MyStaticAppName -g MyRg
+"""
+
+helps['staticapp environments list'] = """
```
Comment (qwordy) [inline src/azure-cli/azure/cli/command_modules/appservice/_help.py:2043]: I think `staticapp environment` is better.

File: src/azure-cli/azure/cli/command_modules/appservice/commands.py [hunk 8]
```diff
@@ -357,3 +359,30 @@ def load_command_table(self, _):
...
+        g.custom_command('list', 'list_staticsites')
+        g.custom_command('browse', 'show_staticsite')
+        g.custom_command('create', 'create_staticsites', supports_no_wait=True)
+        g.custom_command('delete', 'delete_staticsite', supports_no_wait=True, confirmation=True)
+        g.custom_command('disconnect', 'disconnect_staticsite', supports_no_wait=True)
+        g.custom_command('reconnect', 'reconnect_staticsite', supports_no_wait=True)
+
+    with self.command_group('staticapp environments', custom_command_type=staticsite_sdk) as g:
+        g.custom_command('list', 'list_staticsite_environments')
+        g.custom_command('show', 'show_staticsite_environment')
```
Comment (qwordy) [inline src/azure-cli/azure/cli/command_modules/appservice/commands.py:373]:
```suggestion
        g.custom_show_command('show', 'show_staticsite_environment')
```

File: src/azure-cli/azure/cli/command_modules/appservice/static_sites.py [hunk 9]
```diff
@@ -0,0 +1,252 @@
...
+    return sdk_no_wait(no_wait, client.create_or_update_static_site_custom_domain,
+                       resource_group_name=resource_group_name, name=name, domain_name=hostname)
+
+
+def delete_staticsite_domain(cmd, name, hostname, resource_group_name=None, no_wait=False):
+    client = _get_staticsites_client_factory(cmd.cli_ctx)
+    if not resource_group_name:
+        resource_group_name = _get_resource_group_name_of_staticsite(client, name)
+
+    print("After deleting a custom domain, there can be a 15 minute delay for the change to propagate.")
```
Comment (qwordy) [inline src/azure-cli/azure/cli/command_modules/appservice/static_sites.py:85]: Maybe you should use `logger.warning`

File: src/azure-cli/azure/cli/command_modules/appservice/static_sites.py [hunk 10]
```diff
@@ -0,0 +1,252 @@
...
+    if not resource_group_name:
+        resource_group_name = _get_resource_group_name_of_staticsite(client, name)
+
+    return sdk_no_wait(no_wait, client.delete_static_site,
+                       resource_group_name=resource_group_name, name=name)
+
+
+def _parse_pair(pair):
+    list = pair.split("=")
+    return list[0], list[1]
```
Comment (qwordy) [inline src/azure-cli/azure/cli/command_modules/appservice/static_sites.py:212]: Is it possible that "index out of range"

**Comment by qwordy** [comment]
Some suggestions:
1. Write a description for this PR.
2. Pay attention to usage of "a" and plural form in command names and help messages.
3. Write test cases.

**Comment by panchagnula** [comment]
@ElleTojaroon given the create needs a GitHub access token, we need to create mock tests for these. You can look at the existing mock tests for function apps that might help you here.

File: src/azure-cli/azure/cli/command_modules/appservice/_help.py (hunk 2 above)
Comment (ElleTojaroon) [reply src/azure-cli/azure/cli/command_modules/appservice/_help.py:2000]: added

File: src/azure-cli/azure/cli/command_modules/appservice/_help.py (hunk 4 above)
Comment (ElleTojaroon) [reply src/azure-cli/azure/cli/command_modules/appservice/_help.py:2058]: done

File: src/azure-cli/azure/cli/command_modules/appservice/commands.py [hunk 11]
```diff
@@ -358,3 +360,31 @@ def load_command_table(self, _):
         g.custom_command('create', 'create_appserviceenvironment_arm', supports_no_wait=True)
         g.custom_command('update', 'update_appserviceenvironment', supports_no_wait=True)
         g.custom_command('delete', 'delete_appserviceenvironment', supports_no_wait=True, confirmation=True)
+
+    with self.command_group('staticapp', custom_command_type=staticsite_sdk) as g:
```
Comment (panchagnula) [inline src/azure-cli/azure/cli/command_modules/appservice/commands.py:364]: static App commands should be marked Preview.

File: src/azure-cli/azure/cli/command_modules/appservice/utils.py [hunk 12]
```diff
@@ -63,3 +63,9 @@ def call(*args, **kwargs):
                 time.sleep(interval_sec)
         return call
     return decorate
+
+
+def _get_resource_group_from_id(arm_id):
```
Comment (panchagnula) [inline src/azure-cli/azure/cli/command_modules/appservice/utils.py:68]: What is the purpose of this helper? is there a reason we need this specifically for static Apps?
@qwordy does CLI have a shared helper for this?

Comment (qwordy) [reply src/azure-cli/azure/cli/command_modules/appservice/utils.py:68]: from msrestazure.tools import parse_resource_id

Comment (ElleTojaroon) [reply src/azure-cli/azure/cli/command_modules/appservice/utils.py]: removed this helper method and used msrestazure.tool

File: src/azure-cli/azure/cli/command_modules/appservice/_help.py [hunk 13]
```diff
@@ -1991,3 +1991,147 @@
     - name: Delete app service environment.
       text: az appservice ase delete -n MyAseName
 """
+
+helps['staticapp list'] = """
```
[the remark this answers is not in the document]
Comment (ElleTojaroon) [reply src/azure-cli/azure/cli/command_modules/appservice/_help.py]: I updated to staticwebapp, but waiting to see if swa alias is a go or no go.

File: src/azure-cli/azure/cli/command_modules/appservice/_help.py [hunk 14]
```diff
@@ -1991,3 +1991,147 @@
...
+    - name: Disconnect static app.
+      text: az staticapp disconnect -n MyStaticAppName
+"""
+
+helps['staticapp reconnect'] = """
+    type: command
+    short-summary: Connect to a repo and branch following a disconnect command.
+    examples:
+    - name: Connect a repo and branch to static app.
+      text: az staticapp reconnect -n MyStaticAppName --source MyGitHubRepo -b master --token MyAccessToken
```
[the remark this answers is not in the document]
Comment (ElleTojaroon) [reply src/azure-cli/azure/cli/command_modules/appservice/_help.py]: Added the link to get token in help of --token param.

File: src/azure-cli/azure/cli/command_modules/appservice/_help.py [hunk 15]
```diff
@@ -1991,3 +1991,147 @@
...
+    - name: List function app settings of the static app.
+      text: az staticapp appsettings list -n MyStaticAppName
+"""
+
+helps['staticapp appsettings set'] = """
+    type: command
+    short-summary: Set (replace) function app settings of the static app.
+    examples:
+    - name: Set (replace) function app settings of the static app.
+      text: az staticapp appsettings set -n MyStaticAppName --setting-names key1=val1 key2=val2
```
[the remark this answers is not in the document]
Comment (ElleTojaroon) [reply src/azure-cli/azure/cli/command_modules/appservice/_help.py]: Looks like webapps command for app settings also take key value pairs as strings. 

[ for scope in ['appsettings', 'connection-string']:
        with self.argument_context('webapp config ' + scope) as c:
            c.argument('settings', nargs='+', help="space-separated {} in a format of `<name>=<value>`".format(scope))
            c.argument('slot_settings', nargs='+', help="space-separated slot {} in a format of either `<name>=<value>` or `@<json_file>`".format(scope))
            c.argument('setting_names', nargs='+', help="space-separated {} names".format(scope))](https://github.com/Azure/azure-cli/blob/18ac84d464b44a10af5c3d511da32c45bf5f314b/src/azure-cli/azure/cli/command_modules/appservice/_params.py)

File: src/azure-cli/azure/cli/command_modules/appservice/_help.py [hunk 16]
```diff
@@ -1991,3 +1991,147 @@
...
+    - name: Show static app.
+      text: az staticapp show -n MyStaticAppName
+"""
+
+helps['staticapp create'] = """
+    type: command
+    short-summary: Create a static app and returns the app created.
+    examples:
+    - name: Create static app.
+      text: az staticapp create -n MyStaticAppName -g MyRg -s MyGitHubRepo -l Location
```
[the remark this answers is not in the document]
Comment (ElleTojaroon) [reply src/azure-cli/azure/cli/command_modules/appservice/_help.py]: I made --token optional, and throw exception when once isn't passed in. I'll update to make token required.

[the remark this answers is not in the document]
Comment (ElleTojaroon) [reply src/azure-cli/azure/cli/command_modules/appservice/_help.py]: updated

File: src/azure-cli/azure/cli/command_modules/appservice/_help.py [hunk 17]
```diff
@@ -1991,3 +1991,147 @@
...
+    - name: Delete given hostname of the static app.
+      text: az staticapp hostname delete -n MyStaticAppName --hostname HostnameToDelete
+"""
+
+helps['staticapp appsettings list'] = """
+    type: command
+    short-summary: List function app settings of the static app. A function must first be deployed to use this method.
+    examples:
+    - name: List function app settings of the static app.
+      text: az staticapp appsettings list -n MyStaticAppName
```
[the remark this answers is not in the document]
Comment (ElleTojaroon) [reply src/azure-cli/azure/cli/command_modules/appservice/_help.py]: is environment name enough? I made that arg optional with 'default' as default value. 

def list_staticsite_functions(cmd, name, resource_group_name=None, environment_name='default'):

File: src/azure-cli/azure/cli/command_modules/appservice/_help.py [hunk 18]
```diff
@@ -1991,3 +1991,147 @@
...
+    - name: Show static app environment.
+      text: az staticapp environments show -n MyStaticAppName
+"""
+
+helps['staticapp environments functions'] = """
+    type: command
+    short-summary: Show information about functions.
+    examples:
+    - name: Show static app functions.
+      text: az staticapp environments functions -n MyStaticAppName
```
[the remark this answers is not in the document]
Comment (ElleTojaroon) [reply src/azure-cli/azure/cli/command_modules/appservice/_help.py]: is environment name enough? I made that arg optional with 'default' as default value.

def list_staticsite_functions(cmd, name, resource_group_name=None, environment_name='default'):

**Pull request merged**